Fantasy Battle Game/Assets/Scripts/Player/Champions/ChampionsUtilities/Movements/HexMovement.cs

```diff
@@ -139,22 +145,106 @@ public IEnumerator LookAtIEnumerator(Vector3 point)
             orientation_ = transform.localRotation.eulerAngles.y;
         }
 
-        protected void CalculateRoute()
+        private double EstimateCost(Tile current, Tile destination)
         {
+            var dx = Math.Abs(current.Coordinate.FirstCoord - destination.Coordinate.FirstCoord);
+            var dy = Math.Abs(current.Coordinate.SecondCoord - destination.Coordinate.SecondCoord);
+
+            return 1 * (dx + dy) + (Math.Sqrt(2) - 2 * 1) * Math.Min(dx, dy);
+        }
+
+        protected void CalculateRoute(int kindOfAlgorithm)
+        {
+            
             route_.Clear();
-            Tile TempDestinationTile = currentTile;
-            route_.Add(TempDestinationTile);
-            while (TempDestinationTile != DestinationTile)
+                                
+            openList.Clear();
+            closedList.Clear();
+
+            switch (kindOfAlgorithm)
             {
-                var list = SortList(TempDestinationTile, DestinationTile);
+                case 1:
+                    
+                    openList.Add(currentTile);
+                    currentTile.GScore = 0;
+                    currentTile.FScore = EstimateCost(currentTile, DestinationTile);
 
-                var position = 0;
-                while (route_.Contains(list[position]))
-                {
-                    position++;
-                }
-                route_.Add(list[position]);
-                TempDestinationTile = list[position];
+                    while (openList.Any())
+                    {
+                        
+                        var bestTile = openList[0];
+                        
+                        foreach (var elem in openList)
+                        {
+                            if (elem.FScore < bestTile.FScore) bestTile = elem;
+                        }
+
+                        if (bestTile == DestinationTile) break;
+
+                        openList.Remove(bestTile);
+                        closedList.Add(bestTile);
+
+                        List<Tile> neighbours = map_.AvailableNeighbours(bestTile);
+            
+                        foreach (var elem in neighbours)
+                        {
+                            if (closedList.Contains(elem)) continue;
+                            if (!openList.Contains(elem))
+                            {
+                                openList.Add(elem);
+
+                                var gScore = bestTile.GScore + bestTile.Drag / 2 + elem.Drag / 2;
```

Comment:
you make many divisions, without certainity of var being not null!

Reply:
Drag cannot be null. It was initialized probably during map creation.